flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1211,37 +1211,47 @@ struct XEmboxOpConversion : public EmboxCommonConversion<fir::cg::XEmboxOp> {
     mlir::Value zero = genConstantIndex(loc, i64Ty, rewriter, 0);
     mlir::Value one = genConstantIndex(loc, i64Ty, rewriter, 1);
     mlir::Value prevDim = integerCast(loc, rewriter, i64Ty, eleSize);
-    mlir::Value prevPtrOff = one;
+    // To cover all cases (arrays with compile time constant shape, arrays with
+    // dynamic shape, intrinsic types, dependent types like characters), cast
+    // the base to a byte pointer type.  This allows generating a single form of
+    // GEP when shifting the base and keeping this code simple.
```

Comment:
This part of the change (throwing away the type structure) seems like the wrong direction. LLVM's optimizer can do a better job if the IR is correctly and precisely typed. That may be slightly harder here, but was there actually a bug? Converting things to byte arithmetic, when they don't have to be, would be a mistake.

Reply:
Remember too, FIR is target agnostic.

Reply:
> but was there actually a bug?

Yes, took me an hour to understand why some end-to-end test cases were working on some failing for very similar things.

For instance, this was working:
```
subroutine foo(x, i)
  real :: x(i)
  print *, x(11:50:2)
end subroutine
``` 

and this was failing:

```
subroutine foo(x)
  real :: x(100)
  print *, x(11:50:2)
end subroutine
``` 

The issue is that the previous GEP was only OK with dynamic shapes (and constant character length for characters). It was wrong with constant shapes (and dynamic shapes for characters). The reason being the GEP generated was `GEP type, element_shift`, which is ok if type is `i32*` (dynamic shape arrays), but not for constant shape arrays where the type is lowered as `[i32 x100]`. This created a shift of the base  (in element number) of element_shift * 100...
For the characters, that is the reverse, we want to take into account the element length in the shift, so `GEP i8*, element_shift` were wrong.

So, while I agree with you that giving more contexts to LLVM is most likely better, it creates a matrix of cases to handle, which is doable but should be thoroughly tested for all possible type/shape cases. I think we can just live with simple code that cover all cases until we have a stable compiler. It will make our debugging much easier.

I added a TODO to remember that we can improve the generated code.

> Remember too, FIR is target agnostic.

Yes, but at that point we are already filling the descriptor with the element length in bytes, so I think this is not adding any new target dependence aspect.

Reply:
Let's discuss this more. I'm not seeing the connection.